src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/module-splitting.h"
+#include "ir/manipulation.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace ModuleSplitting {
+
+namespace {
+
+std::unique_ptr<Module> initializeSecondary(const Module& primary) {
+  // Create the secondary module and copy trivial properties.
+  auto secondary = std::make_unique<Module>();
+  secondary->features = primary.features;
+  secondary->hasFeaturesSection = primary.hasFeaturesSection;
+  return secondary;
+}
+
+void shareImportableItems(Module& primary,
+                          Module& secondary,
+                          Name importNamespace) {
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
+      exports[ex->value] = ex->name;
+    }
+  }
+
+  auto makeImportExport = [&](Importable& primaryItem,
+                              Importable& secondaryItem,
+                              Name exportName,
+                              ExternalKind kind) {
+    secondaryItem.name = primaryItem.name;
+    secondaryItem.hasExplicitName = primaryItem.hasExplicitName;
+    secondaryItem.module = importNamespace;
+    auto exportIt = exports.find(primaryItem.name);
+    if (exportIt != exports.end()) {
+      secondaryItem.base = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, exportName);
+      primary.addExport(new Export{exportName, primaryItem.name, kind});
+      secondaryItem.base = exportName;
+    }
+  };
+
+  // TODO: Be more selective by only sharing global items that are actually used
+  // in the secondary module, just like we do for functions.
+
+  auto shareMemory = [&](Memory* memory) {
+    secondary.memory.exists = true;
+    secondary.memory.initial = memory->initial;
+    secondary.memory.max = memory->max;
+    secondary.memory.shared = memory->shared;
+    secondary.memory.indexType = memory->indexType;
+    makeImportExport(*memory, secondary.memory, "memory", ExternalKind::Memory);
+  };
+  ModuleUtils::iterImportedMemories(primary, shareMemory);
+  ModuleUtils::iterDefinedMemories(primary, shareMemory);
+
+  auto shareTable = [&](Table* table) {
+    secondary.table.exists = true;
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
+  ModuleUtils::iterImportedTables(primary, shareTable);
+  ModuleUtils::iterDefinedTables(primary, shareTable);
+
+  auto shareGlobal = [&](Global* global) {
+    assert(primary.features.hasMutableGlobals() &&
+           "TODO: add wrapper functions for disallowed mutable globals");
```

Comment:
Since the globals are exported and imported, we need the mutable-globals feature to be enabled if there are any mutable globals in the module. I guess we technically don't need this assert for modules that don't have any mutable globals, but that seems niche.

Reply:
Sorry I'm not sure if I understand. If a primary module doesn't have mutable globals, both primary and secondary modules will not need them, so everything is fine. If the primary module originally has mutable globals enabled, the secondary module will copy all those features, and both of them will have mutable globals, and everything's fine here. Why do we need to require anything about mutable globals for module splitting? Why is the mutable globals feature treated specially unlike all other features?

Reply:
The `mutable-globals` feature is a bit of a misnomer. What this feature actually allows is _imports and exports_ of mutable globals. It's totally valid to have mutable globals without it in MVP, but not to import or export them. So we need to check this feature here because importing or exporting a mutable global that was not previously imported or exported introduces a new use of the `mutable-globals` feature that was not there before.

Reply:
I see... I was seriously confused about what mutable globals feature does. Thanks for pointing that out.

Reply:
Yeah the feature name really gives the wrong idea 😅